### DISCUSSION TITLE
ImageSharp 3.0 global configuration for decoders

### DISCUSSION THREAD

**Discussion by hey-red** [SixLabors/ImageSharp#2382] [Q&A] [answered]
In 2.x version we could set a global configuration for decoders:

```C#
configuration.ImageFormatsManager.SetDecoder(GifFormat.Instance, new GifDecoder()
{
       DecodingMode = FrameDecodingMode.First,
});
```

How to achieve this in 3.0? Or the only way is to use a decoder instance?

```C#
var image = JpegDecoder.Instance.Decode(new JpegDecoderOptions { ResizeMode = JpegDecoderResizeMode.ScaleOnly }, input);
```

**Comment by tocsoft** [comment, marked as answer]
What are you seeing that implies your first code snippet doesn't work?

**Reply by hey-red** [reply]
As I can see, the general options has been moved from decoders to separate type https://github.com/SixLabors/ImageSharp/discussions/2091

But what about for global Configuration? In 2.x I can configure decoders specific options and then pass configuration into Image.Load method.

**Reply by JimBobSquarePants** [reply]
You can pass that configuration via the `DecoderOptions`. It has a `Configuration` property that defaults to the default shared instance.

**Reply by hey-red** [reply]
@JimBobSquarePants Yes, I see, but I don't understand how do this with [JpegDecoderOptions.ResizeMode](https://github.com/SixLabors/ImageSharp/blob/main/src/ImageSharp/Formats/Jpeg/JpegDecoderOptions.cs#L17)

All options that implements `ISpecializedDecoderOptions` can be used only with decoder instance, and not possible to set it on global configuration, right?

**Reply by JimBobSquarePants** [reply]
Yes. The general options are for general use cases. I.E. `Image.Load` etc.

Using the decoders directly is an advanced process not recommended for general use. 

You should only use the decoder directly if you require specialist options and only if you are 100% confident that you know the type of image you are decoding. (_I’ve soon far too many people complain because they’ve tried to decode file with the wrong extension directly with a decoder_) 

You can use the `Image.DetectFormat` APIs to identify unknown types before choosing the correct decoder if you do not know it.

You’ve not lost functionality. You’ve gained new functionality.  

The API redesign is specifically implemented to provide guidance as to the correct path to choose to avoid common foot guns.

**Reply by hey-red** [reply]
@JimBobSquarePants Understood. Looks reasonable.
The one problem that I can see that ImageSharp.Web(v2) does't provide an alternative api for this use case. 
Thanks for help.

**Reply by JimBobSquarePants** [reply]
Yet… Still working on the next release of that library.
